File: src/test/java/plugin/utilities/ASTNodePrinter.java
package plugin.utilities;

import com.intellij.lang.ASTNode;
import com.intellij.psi.tree.TokenSet;

public class ASTNodePrinter {

    public static String printDirectoryTree(ASTNode folder) {
        int indent = 0;
        StringBuilder sb = new StringBuilder();
        printDirectoryTree(folder, indent, sb);
        return sb.toString();
    }

    private static void printDirectoryTree(ASTNode folder, int indent, StringBuilder sb) {
        sb.append(getIndentString(indent));
        sb.append(toPrettyName(folder.getPsi().getNode().getElementType().toString()));
        sb.append("\n");
        for (ASTNode file : folder.getChildren(TokenSet.ANY))
            if (file.getChildren(TokenSet.ANY).length > 0)
                printDirectoryTree(file, indent + 1, sb);
            else
                printFile(file.getText(), indent + 1, sb);
    }

    private static void printFile(String name, int indent, StringBuilder sb) {
        sb.append(getIndentString(indent));
        sb.append(toPrettyName(name));
        sb.append("\n");
    }

    private static String getIndentString(int indent) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < indent; i++)
            sb.append("   ");
        return sb.toString();
    }

    private static String toPrettyName(String name) {
        if (name.length() == 1) {
            char c = name.charAt(0);
            if (Character.isWhitespace(c))
                return Character.getName(c).toLowerCase();
        }
        return name;
    }

}

File: src/test/kotlin/plugin/ASTNodePrinter.kt
package plugin

import com.intellij.lang.ASTNode
import com.intellij.psi.tree.TokenSet

object ASTNodePrinter {

    fun printDirectoryTree(folder: ASTNode): String {
        val indent = 0
        val sb = StringBuilder()
        printDirectoryTree(folder, indent, sb)
        return sb.toString()
    }

    private fun printDirectoryTree(folder: ASTNode, indent: Int, sb: StringBuilder) {
        sb.append(getIndentString(indent))
        sb.append(folder.psi.toString())
        sb.append("\n")
        for (file in folder.getChildren(TokenSet.ANY))
            if (file.getChildren(TokenSet.ANY).isNotEmpty())
                printDirectoryTree(file, indent + 1, sb)
            else
                printFile(file.toString(), indent + 1, sb)
    }

    private fun printFile(name: String, indent: Int, sb: StringBuilder) {
        sb.append(getIndentString(indent))
        sb.append(name)
        sb.append("\n")
    }

    private fun getIndentString(indent: Int) = "  ".repeat(indent)

}
